pennylane/fourier/reconstruct.py

```diff
@@ -0,0 +1,295 @@
+    "function to be reconstructed."
+)
+
+
+def _reconstruct_gen(fun, spectrum, shifts=None, fun_at_zero=None):
+    r"""Reconstruct a univariate (real-valued) Fourier series with given spectrum.
+    Args:
+        fun (callable): Fourier series to reconstruct with signature ``float -> float``
+        spectrum (Sequence): Frequency spectrum of the Fourier series; non-positive
+            frequencies are ignored
+        shifts (list): Shift angles at which to evaluate ``fun`` for the reconstruction
+            Chosen equidistantly within the interval :math:`[0, 2\pi/f_\text{min}]` if ``shifts=None``
+            where :math:`f_\text{min}` is the smallest frequency in ``spectrum``.
+        fun_at_zero (float): Value of ``fun`` at zero. If :math:`0` is among the ``shifts``
+            and ``fun_at_zero`` is provided, one evaluation of ``fun`` is saved.
+
+    Returns:
+        callable: Reconstructed Fourier series with :math:`R` frequencies in ``spectrum``,
+        as ``qml.numpy`` based function and coinciding with ``fun`` on :math:`2R+1` points.
+    """
+    # pylint: disable=unused-argument
+
+    have_fun_at_zero = fun_at_zero is not None
+    have_shifts = shifts is not None
+    # For an empty/trivial spectrum, the function simply is constant
+    if spectrum in ([], [0.0]):
+        if have_shifts:
+            fun_value = fun(shifts[0])
+            if have_fun_at_zero:
+                warnings.warn(_warn_text_fun_at_zero_ignored)
+        else:
+            fun_value = fun_at_zero if have_fun_at_zero else fun(0.0)
+
+        def constant_fn(x):
+            """Univariate reconstruction of a constant Fourier series."""
+            return fun_value
+
+        return constant_fn
+
+    spectrum = np.array([f for f in spectrum if f > 0.0])
+    f_max = max(spectrum)
+
+    # If no shifts are provided, choose equidistant ones
+    if not have_shifts:
+        R = len(spectrum)
+        shifts = np.arange(-R, R + 1) * 2 * np.pi / (f_max * (2 * R + 1)) * R
+        zero_idx = R
+        need_fun_at_zero = True
+    elif have_fun_at_zero:
+        zero_idx = np.where(np.isclose(shifts, 0.0))[0]
+        zero_idx = zero_idx[0] if len(zero_idx) > 0 else None
+        need_fun_at_zero = zero_idx is not None
+
+    # Take care of shifts close to zero if fun_at_zero was provided
+    if have_fun_at_zero and need_fun_at_zero:
+        # Only one shift may be zero at a time
+        shifts = np.concatenate([[shifts[zero_idx]], shifts[:zero_idx], shifts[zero_idx + 1 :]])
+        evals = np.array([fun_at_zero] + list(map(fun, shifts[1:])))
+    else:
+        if have_fun_at_zero and not need_fun_at_zero:
+            warnings.warn(_warn_text_fun_at_zero_ignored)
+        evals = np.array(list(map(fun, shifts)))
+
+    L = len(shifts)
+    # Construct the coefficient matrix case by case
+    C1 = np.ones((L, 1))
+    C2 = np.cos(np.outer(shifts, spectrum))
+    C3 = np.sin(np.outer(shifts, spectrum))
+    C = np.hstack([C1, C2, C3])
```

Comment:
ah, just realized this whole function is NumPy only :)

Reply:
I meant to update this, sorry!